### ISSUE TITLE
Questions + an issue

### ISSUE THREAD

**Issue by bblodfon** [ropensci/aorsf#16]
Hi @bcjaeger, I was checking your package via the `mlr3extralearners` (really good work!!!) and had some questions/notes for you when you have some time:

1. I see no parameter like `num.threads` in `ranger`, so I guess parallelization is not supported? Is it because its difficult to implement or somehow it is not possible in case of ORSFs?
2. I was thinking of tuning the `control_type` (`fast` vs `cph` vs `net`). Do you think that it would make sense to group some of these as a **different random forest learner** as it might be the case that e.g. `fast` and `cph` produce similar forests/results/predictions (using the same dataset and using `control_cph_iter_max = to_tune(1,20)` for example) compared to `net` (where tuning the `alpha` might result in completely different solutions). In your recent arXiv paper you had each as a separate learner (e.g. `aorsf-net`, `aorsf-fast` and `aorsf-cph`). Or you would advise in using `fast` in all cases (i.e. not tune at all) based on the results from your recent arXiv paper (though `alpha` wasn't tuned there if I understood correctly)?
3. In `orsf_control_net`, do you leave the tuning of the `lambda` to glmnet? e.g. is `cv.glmnet` being used "under the hood" :) ? (that would explain the reason it takes so much more time)
4. For tuning purposes, is the `split_min_obs`/`split_min_events` the equivalent of `min.node.size` in `ranger`? i.e. a node should have **at least** `split_min_obs` observations (out of which `split_min_events` are events), to consider splitting it?
5. Have you tested the package in more high-dim settings (in the paper only one dataset was with `n < p`)? I have a dataset of 145 observations x 10000 features and there seems to be some overhead before the training phase (before `growing tree No.` starts appearing) and during prediction (maybe has to do with not having parallelized trees?). Especially, `permutation` importance (which I want to use in my [implementation of RFE + RSFs](https://github.com/bblodfon/survmob)) takes so much more time to compute (only used `100` trees, stopped it after some minutes...)! With `ranger`, even with no parallelization, I can train and predict with the same dataset in less than 3 secs. Let me know if you want me to share the dataset for more investigation, it would be great if somehow things could be sped up a bit!
6. In the arXiv paper, you mention the '1 point region of practical equivalence' for comparing the different learners using a Bayesian linear mixed model (awesome!). Is that 1 point difference equivalent to a 0.01 difference in the C-index for example since everything is scaled by 100? Is that is the case, is even a 5 point average difference a practical difference at all?
7. A small issue I found (maybe for `mlr3extralearners`):
``` r
library(mlr3proba)
#> Loading required package: mlr3
library(mlr3extralearners)

task_mRNA = readRDS(file = gzcon(url('https://github.com/bblodfon/paad-survival-bench/blob/main/data/task_mRNA_flt.rds?raw=True'))) # 1000 features

dsplit = mlr3::partition(task_mRNA, ratio = 0.8)
train_indxs = dsplit$train
test_indxs  = dsplit$test

orsf = lrn('surv.aorsf',
  importance = 'none',
  oobag_pred_type = 'surv',
  attach_data = FALSE,
  verbose_progress = TRUE, 
  n_tree = 10
)

orsf$train(task = task_mRNA, row_ids = train_indxs)
#>  growing tree no. 9 of 10
p = orsf$predict(task = task_mRNA, row_ids = test_indxs)
#> Error: Assertion on 'length(times)' failed: FALSE.
```
 
<sup>Created on 2023-02-27 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

I don't want to carry the training `data` to calculate importance or other things so I used `attach_data = FALSE`, but it seems in the [mlr3extralearners wrapper code](https://github.com/mlr-org/mlr3extralearners/blob/main/R/learner_aorsf_surv_aorsf.R#L164) you use the `$model$data` slot **for prediction either way** which results in the above error. 

So I guess since the `time` and `status` of the training data are always needed for prediction and we would like to keep the `attach_data` option but allow prediction to work when `attach_data = FALSE`, then maybe consider attaching these two into the `$model` slot and accessing them on the wrapper code (rather than the `model$data`)?

**Comment by bcjaeger** [comment]
Hi @bblodfon, thank you!! I am thrilled this software is getting used. I'm also happy to discuss these questions and will also look into the issue you've found. I am traveling this week but will hopefully have time to review the code next week.

> I see no parameter like num.threads in ranger, so I guess parallelization is not supported? Is it because its difficult to implement or somehow it is not possible in case of ORSFs?

Correct, parallel support isn't there yet. I regret not coding it that way in the first place. I plan to implement it in the future, but am currently focused on getting grant support for continuing to develop `aorsf`. I.e., parallel computing support is a ways off.

> I was thinking of tuning the control_type (fast vs cph vs net). Do you think that it would make sense to group some of these as a different random forest learner as it might be the case that e.g. fast and cph produce similar forests/results/predictions (using the same dataset and using control_cph_iter_max = to_tune(1,20) for example) compared to net (where tuning the alpha might result in completely different solutions). In your recent arXiv paper you had each as a separate learner (e.g. aorsf-net, aorsf-fast and aorsf-cph). Or you would advise in using fast in all cases (i.e. not tune at all) based on the results from your recent arXiv paper (though alpha wasn't tuned there if I understood correctly)?

If you have the time, it certainly wouldn't hurt to think `control_type` as a tuning parameter. Looking at the individual datasets where we ran benchmarks in the arxiv paper, I do see some cases where one `orsf` method outperformed the others (although they were very similar overall). 

> In orsf_control_net, do you leave the tuning of the lambda to glmnet? e.g. is cv.glmnet being used "under the hood" :) ? (that would explain the reason it takes so much more time)

Great question. What is currently used is this R function: https://github.com/ropensci/aorsf/blob/main/R/penalized_cph.R. For computational efficiency, we just run glmnet and then use the solution that has the largest degrees of freedom beneath a maximum value. I would guess that cv.glmnet could eek out a little more prediction accuracy compared to the version we currently use, with the obvious trade-off of computation time. Since `orsf` lets you pass your own function to make linear combinations of variables, you _should_ be able to pass in your own function using `cv.glmnet`. However, it may have some trouble running with smaller samples. Let me know how it goes if you do!

> For tuning purposes, is the split_min_obs/split_min_events the equivalent of min.node.size in ranger? i.e. a node should have at least split_min_obs observations (out of which split_min_events are events), to consider splitting it?

I think `ranger`'s `min.node.size` corresponds to `orsf`'s `leaf_min_obs`, since `leaf_min_obs` is the minimum number of observations in a leaf node of the tree 

> Have you tested the package in more high-dim settings (in the paper only one dataset was with n < p)? I have a dataset of 145 observations x 10000 features and there seems to be some overhead before the training phase (before growing tree No. starts appearing) and during prediction (maybe has to do with not having parallelized trees?). Especially, permutation importance (which I want to use in my [implementation of RFE + RSFs](https://github.com/bblodfon/survmob)) takes so much more time to compute (only used 100 trees, stopped it after some minutes...)! With ranger, even with no parallelization, I can train and predict with the same dataset in less than 3 secs. Let me know if you want me to share the dataset for more investigation, it would be great if somehow things could be sped up a bit!


First, that's super impressive - `ranger` is quite good. 
As far as `orsf` goes, there are a number of checks done prior to model fitting that can take some time with higher dimensional data. E.g., checking for invalid variable types, checking for NaN values, and checking for missing values. It's worth doing these checks though - they've already saved me a few times. I certainly did not optimize `aorsf` for variable importance as much as I did for tree growing and prediction. Definitely open to improving computation here in the future though. Out of curiousity, did you try ANOVA importance?

> In the arXiv paper, you mention the '1 point region of practical equivalence' for comparing the different learners using a Bayesian linear mixed model (awesome!). Is that 1 point difference equivalent to a 0.01 difference in the C-index for example since everything is scaled by 100? Is that is the case, is even a 5 point average difference a practical difference at all?

I'm glad you liked the modeling approach! You're correct about the 1 point difference being a difference of 0.01. I do not know what a real practical difference is when it comes to the C-statistic, but generally if a model can improve by 1 point (e.g., increase from 0.75 to 0.76), I think of that as helpful. Going from 0.75 to 0.80 just by switching models would seem like a home-run to me, but that's just my personal opinion. Do you think a 1 point region was too narrow for the C-statistic?

**Comment by bblodfon** [comment]
If `mtry` changes in each node to satisfy such a condition (and the more deep grows the tree, `mtry` becomes smaller as there are less and less events), then does it make sense to have it tuned, what do you think?

Again refering only to the `fast` and `cph` control methods, do you have any other suggestions for parameters to tune apart from `n_tree` and `leaf_min_obs`?

**Comment by bcjaeger** [comment]
> If mtry changes in each node to satisfy such a condition (and the more deep grows the tree, mtry becomes smaller as there are less and less events), then does it make sense to have it tuned, what do you think?

That would be an interesting tuning parameter, but I am not sure it would make a big difference. 

> Again refering only to the fast and cph control methods, do you have any other suggestions for parameters to tune apart from n_tree and leaf_min_obs?

I would not suggest tuning `n_tree`, as random forests generally do not overfit with the addition of more trees. Just make sure it's high enough - 500 usually suffices but going up to 5,000 wouldn't hurt. 

`leaf_min_obs` is the main thing to tune. You may also try tuning `n_retry`. Higher values of this parameter may help when there isn't much signal in the data.
